apps/cli/lib/parser/mana_parser.ex

```diff
@@ -0,0 +1,133 @@
+defmodule CLI.Parser.ManaParser do
+  @moduledoc """
+  Parses for command line arguments for the CLI.
+  """
+
+  @type sync_arg_keywords :: [provider: String.t(), provider_url: String.t()]
+
+  @default_chain_name "ropsten"
+  @default_no_discovery false
+  @default_no_sync false
+  @default_bootnodes "from_chain"
+
+  @doc """
+  Parsers args for syncing
+
+  ## Options:
+    * `--chain` - Chain to load data from (default: ropsten)
+    * `--no-discovery` - Perform discovery (default: false)
+    * `--no-sync` - Perform syncing (default: false)
+    * `--bootnodes` - Comma separated list of bootnodes (default: from_chain)
+
+  ## Examples
+
+      iex> CLI.Parser.ManaParser.mana_args(["--chain", "ropsten", "--no-discovery", "--no-sync"])
+      {:ok, %{
+        chain_name: :ropsten,
+        discovery: false,
+        sync: false,
+        bootnodes: :from_chain
+      }}
+
+      iex> CLI.Parser.ManaParser.mana_args(["--chain", "foundation", "--bootnodes", "enode://google.com,enode://apple.com"])
+      {:ok, %{
+        chain_name: :foundation,
+        discovery: true,
+        sync: true,
+        bootnodes: ["enode://google.com", "enode://apple.com"]
+      }}
+
+      iex> CLI.Parser.ManaParser.mana_args([])
+      {:ok, %{
+        chain_name: :ropsten,
+        discovery: true,
+        sync: true,
+        bootnodes: :from_chain
+      }}
+
+      iex> CLI.Parser.ManaParser.mana_args(["--chain", "pony"])
+      {:error, "Invalid chain: pony"}
+  """
+  @spec mana_args([String.t()]) ::
+          {:ok,
+           %{
+             chain_name: atom(),
+             discovery: boolean(),
+             sync: boolean(),
+             bootnodes: :from_chain | list(String.t())
+           }}
+          | {:error, String.t()}
+  def mana_args(args) do
+    {kw_args, _extra} =
+      OptionParser.parse!(args,
+        switches: [chain: :string, no_discovery: :boolean, no_sync: :boolean, bootnodes: :string]
+      )
+
+    with {:ok, chain_name} <- get_chain_name(kw_args) do
+      with {:ok, discovery} <- get_discovery(kw_args) do
```

Comment:
why so many withs?

Reply:
I think this could be written as:

```ex
with {:ok, chain_name} <- get_chain_name(kw_args),
     {:ok, discovery} <- get_discovery(kw_args),
     {:ok, sync} <- get_sync(kw_args),
     {:ok, bootnodes} <- get_bootnodes(kw_args) do
     {
       :ok,
       %{
         chain_name: chain_name,
         discovery: discovery,
         sync: sync,
         bootnodes: bootnodes
       }
    }
end
```

If any of the first 4 lines return a non `{:ok, _}` result then it will short circuit and return the "errored" value. If all successfully return an `{:ok, _)` value then the final result will be the `do`.